### PR TITLE
Completed preliminary parse tree to AST mapper

### PR DESCRIPTION
**Pull request by jenrik**
Added terminator to constant declaration.
Named production parts for ease of use.
Collapsed expressions. ANTLR 4 automaticly does precedence based on
declaration order.
Updated mapper to use named production parts.
Added missing modulo expression mapping.
Added missing function expression mapping.
Added state index expression mapping.
Added missing array literal expression mapping.
Added missing if statement mapping.
Added missing return statement mapping.
Added missing neighbourhood declaration mapping.
Added function declaration mapping.
Added syntax stress example.

### CODE REVIEW

File: compiler/src/main/antlr/Cellmata.g4 [hunk 1]
```diff
@@ -93,37 +100,37 @@ bool_literal
...
+    | left=expr OP_DIVIDE right=expr # divisionExpr
+    | left=expr OP_MODULO right=expr # moduloExpr
+    | OP_INCREMENT value=expr # preIncExpr
+    | OP_DECREMENT value=expr # preDecExpr
+    | OP_PLUS value=expr # positiveExpr
+    | OP_MINUS value=expr # negativeExpr
+    | OP_NOT value=expr # inverseExpr
+    | value=expr OP_INCREMENT # postIncExpr
+    | value=expr OP_DECREMENT # postDecExpr
+    | value=expr SQ_BRACKET_START DIGITS SQ_BRACKET_END # arrayLookupExpr
```
Comment (NicEastvillage) [inline compiler/src/main/antlr/Cellmata.g4]:
```suggestion
    | value=expr SQ_BRACKET_START expr SQ_BRACKET_END # arrayLookupExpr
```

File: compiler/src/main/kotlin/ast/ast.kt [hunk 2]
```diff
@@ -76,27 +75,22 @@ data class InverseExpr(val ctx: ParseTree, val value: Expr) : Expr()
 
 data class ArrayLookupExpr(val ctx: ParseTree, val ident: String = MAGIC_UNDEFINED_STRING, val index: Int = -1) : Expr()
 
-data class ArrayBodyExpr(val ctx: ParseTree, val type: String?, val values: List<Expr>) : Expr()
+data class ArrayBodyExpr(val ctx: ParseTree, val type: String? = MAGIC_UNDEFINED_STRING, val values: List<Expr>) : Expr()
```
Comment (NicEastvillage) [inline compiler/src/main/kotlin/ast/ast.kt]: I think we should make types to an enum.

File: compiler/src/main/kotlin/ast/ast.kt [hunk 3]
```diff
@@ -114,14 +108,23 @@ data class StateDecl(
     val body: List<Stmt>
 ) : Decl()
 
-data class Coordinate(val axes: List<Int>)
+data class Coordinate(val ctx: ParseTree, val axes: List<Int> = emptyList())
```
Comment (NicEastvillage) [inline compiler/src/main/kotlin/ast/ast.kt:114]:
```suggestion
data class Coordinate(val ctx: ParseTree, val coordX: Int, val coordY: Int?)
```

File: compiler/src/main/kotlin/ast/mapper.kt [hunk 4]
```diff
@@ -109,34 +114,55 @@ private fun visitExpr(node: ParseTree): Expr {
...
+                        expr = visitExpr(it.if_stmt_block().if_stmt_condition().expr()),
+                        block = visitCodeBlock(it.if_stmt_block().code_block())
+                    )
+                }
+            ).flatten(),
+            elseBlock = when (node.if_stmt_else()) {
+                // null???
+                is CellmataParser.If_stmt_elseContext -> visitCodeBlock(node.if_stmt_else().code_block())
+                else -> throw AssertionError("Unexpected tree node")
+            }
```
Comment (NicEastvillage) [inline compiler/src/main/kotlin/ast/mapper.kt:163]: Here will be problems when the else-case is not present as the comment `// nul???` also suggests

File: compiler/src/main/kotlin/ast/mapper.kt [hunk 5]
```diff
@@ -146,38 +172,37 @@ private fun visitDecl(node: ParseTree): Decl {
...
     }
 }
 
+fun visitCodeBlock(block: CellmataParser.Code_blockContext): List<Stmt> {
+    return block.children
+        .filter { it !is TerminalNode } // Remove terminals
+        .filter { !(it is CellmataParser.StmtContext && it.childCount == 0) }
+        .filter { !(it is CellmataParser.StmtContext && it.getChild(0) is TerminalNode ) }
+        .map { visitStmt(it) }
+}
```
Comment (NicEastvillage) [inline compiler/src/main/kotlin/ast/mapper.kt:202]: This is weird. What is this filtering for? Why isn't children just 0 or more statements?

File: compiler/src/main/kotlin/main.kt [hunk 6]
```diff
@@ -22,4 +16,6 @@ fun main() {
     val ast = visit(startContext)
 
     println(ast)
+
+    listOf<Unit>()
```
Comment (NicEastvillage) [inline compiler/src/main/kotlin/main.kt]: Why??

**Review by NicEastvillage — CHANGES_REQUESTED** [review]
A few things seems a bit weird. Check those please.

File: compiler/src/main/kotlin/ast/ast.kt (hunk 3 above)
Comment (jenrik) [reply compiler/src/main/kotlin/ast/ast.kt:114]: Having the axes as a list allows for easier data manipulation and validation, and have proven useful to me

File: compiler/src/main/kotlin/ast/ast.kt (hunk 2 above)
Comment (jenrik) [reply compiler/src/main/kotlin/ast/ast.kt]: I'm not sure enums are the right solution, how would you represent an array of integers?

**Comment by jenrik** [comment]
The ability to use a semicolon as a no-op statement has been removed

**Review by NicEastvillage — APPROVED** [review]

**Pull request merged**